### ISSUE TITLE
[TCK Challenge] SerializersCustomizationCDITest assumes different behaviour than non-CDI SerializersCustomizationTest

### ISSUE THREAD

**Issue by struberg** [jakartaee/jsonb-api#344]
It seems that the deserialiser in SerializersCustomizationCDITest is expecting to see a START_OBJECT of the actual items instead of the actual current JsonValue event which is a START_ARRAY.

https://github.com/jakartaee/jsonb-api/blob/master/tck/src/main/java/ee/jakarta/tck/json/bind/cdi/customizedmapping/serializers/model/serializer/AnimalListDeserializerInjected.java#L44

This is different to the (imo correct) implementation in the deserialiser of the non-CDI TCK test which expects to see the START_ARRAY.

https://github.com/jakartaee/jsonb-api/blob/master/tck/src/main/java/ee/jakarta/tck/json/bind/customizedmapping/serializers/model/serializer/AnimalListDeserializer.java#L40

The reason why I think that START_ARRAY is correct is that only this way we can support more complex scenarios.
The current value to parse is 
```
{ "animals" : [ { "type" : "cat", "cuddly" : true, "age" : 5, "furry" : true, "name" : "Garfield" , "weight" : 10.5}, { "type" : "dog", "barking" : true, "age" : 3, "furry" : false, "name" : "Milo", "weight" : 5.5}, { "type" : "animal", "age" : 6, "furry" : false, "name" : "Tweety", "weight" : 0.5} ] }
```

But imagine having a more complex JSON:

```
{
"animals": 
	{
	"size":3,
	"values": [ { "type" : "cat", "cuddly" : true, "age" : 5, "furry" : true, "name" : "Garfield" , "weight" : 10.5}, { "type" : "dog", "barking" : true, "age" : 3, "furry" : false, "name" : "Milo", "weight" : 5.5}, { "type" : "animal", "age" : 6, "furry" : false, "name" : "Tweety", "weight" : 0.5} ]
	}
}
```

For this case the jsonParser.next() should imo point at the START_OBJECT just before the 'size' attribute, right?

**Comment by jeanouii** [comment]
Yeah, the behavior should be the same whereas you run on CDI environment or plain Java SE mode.

To find the correct behavior, the specification might help or the related part of the RFC mentioned in the spec.

**Comment by jbescos** [comment]
@struberg note that for the class:
```
public class AnimalShelterWithSerializer {
    @JsonbTypeSerializer(AnimalListSerializer.class)
    @JsonbTypeDeserializer(AnimalListDeserializer.class)
    private List<Animal> animals = new ArrayList<>();
```
`AnimalListDeserializer` will deserialize only `animals` list. It will not deserialize the class `AnimalShelterWithSerializer`.

You can add the field `size` in that class with no problems, and it will not interfere with the `AnimalListDeserializer`.

I agree with `the behavior should be the same whereas you run on CDI environment or plain Java SE mode`, but looking at your 2 links, in both cases there is `JsonParser.Event.START_OBJECT`.

**Comment by jbescos** [comment]
@struberg one question, you said that:

```
This is different to the (imo correct) implementation in the deserialiser of the non-CDI TCK test which expects to see the START_ARRAY.

https://github.com/jakartaee/jsonb-api/blob/master/tck/src/main/java/ee/jakarta/tck/json/bind/customizedmapping/serializers/model/serializer/AnimalListDeserializer.java#L40
```

But in the link I see `JsonParser.Event.START_OBJECT`. So both (SE and CDI) are consistent.

**Comment by struberg** [comment]
hi! The difference is that the non-CDI version scans over the events UNTIL it hits START_OBJECT. 

```
        while (jsonParser.hasNext()) {
            JsonParser.Event event = jsonParser.next();
            while (event == JsonParser.Event.START_OBJECT) {
                animals.add(animalDeserializer.deserialize(jsonParser,
                                                           deserializationContext, type));
                event = jsonParser.next();
            }
        }
```
Which is imo correct as the beginning of the json content for the "animals" attribute is a START_ARRAY.

This is fundamentally different to the CDI version which expects START_OBJECT as first event.

**Comment by jbescos** [comment]
Thanks for clarification @struberg. Basically you want to have the same logic we use in `AnimalListDeserializer` for CDI.

I agree, although the JSON comes from the TCK test, so I don't expect it is going to change.

**Comment by rmannibucau** [comment]
+1, if CDI and standalone tests are not 1-1 (at injection exception) it is a TCK bug which must be fixed in next TCK delivery and test must be excluded in between. An implementation passing this does not respect the spec which expects both to behave the same JSON wide.

@struberg do you want to PR the fix?

**Comment by struberg** [comment]
> Thanks for clarification @struberg. Basically you want to have the same logic we use in `AnimalListDeserializer` for CDI.

I want to have the same logic we have in the non-CDI AnimalListDeserializer also in the CDI version. Because I honestly believe that START_OBJECT is plain wrong. The Object which gets converted is the json array. So the correct event is START_ARRAY. Having it the same like in the non-CDI version at least makes the TCK pass for both impls. I'll ship a PR which allows both.

**Comment by struberg** [comment]
What happened to this challenge?

**Comment by jbescos** [comment]
@struberg this is not the official repository of TCKs. We just copy-paste the tests so we eagerly know whether we are breaking the spec when we make any commit. I think I didn't express it clearly in [this comment](https://github.com/jakartaee/jsonb-api/issues/344#issuecomment-1732943814).

Could you forward this issue here, please?: https://github.com/jakartaee/platform-tck

**Comment by struberg** [comment]
It that's the case that the TCK doesn't get maintained over here anymore, then we should please delete the whole directory and add the link to the proper location to the readme!

**Comment by struberg** [comment]
And back again: the platform spec EXPLICITLY SAYS: "TCK Challenges MUST be filed via the *specification project’s* issue tracker using the label challenge and include the following information:"

The "specification project" is this very project here! I read this as that we still handle it the way we've done this the last 20 years: the EE umbrella project only is for umbrella challenges, and the specs handle their own challenges themselves. Is this interpretation correct @m0mus ?

**Comment by m0mus** [comment]
@struberg yes, you are right.

**Comment by jbescos** [comment]
Ok, I think the way to go is to create a PR here https://github.com/jakartaee/platform-tck/tree/main/jsonp with the link of this TCK-Challenge. And if it gets accepted, we will bring it here too. Otherwise I am afraid we could have these tests out of sync. (Imagine the case we merge it here first, and it is rejected in https://github.com/jakartaee/platform-tck/tree/main/jsonp).

Are you going to create the PR @struberg?, or do you want me to create it?.

**Comment by jungm** [comment]
Any chance we can get this challenge accepted and resolved? Its one of the two excludes we carry on right now in apache johnzon that blocks us from actually being certified

Sorry for the ping @KyleAure but i've seen you active here the last couple weeks and maybe we can finally get this done

**Comment by KyleAure** [comment]
@struberg -- I read through this issue and it seems like there is concensous here that the CDI and non-CDI tests should have parity in behavior.  I have created a PR in our upstream branch to propose a fix here: https://github.com/jakartaee/jsonb-api/pull/413

Let me know if this is the fix you are looking for or if I misinterpreted the issue.
If it is I can back port the fix to the 3.0.x stream and work on doing a service release of the TCK for you to certify against (or disable the test but I'd rather fix it downstream)
